Microsoft.Identity.Web/Client/TokenAcquisition.cs

```diff
@@ -155,110 +137,55 @@ public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceiv
-
-            // Case of a lazy OBO
-            Claim jwtClaim = context.User.FindFirst("jwt");
-            if (jwtClaim != null)
+            try
             {
-                (context.User.Identity as ClaimsIdentity).RemoveClaim(jwtClaim);
-                var result = await application.AcquireTokenOnBehalfOf(scopes.Except(scopesRequestedByMsalNet), new UserAssertion(jwtClaim.Value))
-                                        .ExecuteAsync();
+                // Use MSAL to get the right token to call the API
+                var application = GetOrBuildConfidentialClientApplication(context, context.User);
+
+                var accounts = await application.GetAccountsAsync();
+
+                var result = await application.AcquireTokenSilent(
+                    scopes.Except(scopesRequestedByMsalNet),
+                    accounts.FirstOrDefault())
+                    .ExecuteAsync();
+
                 return result.AccessToken;
             }
-            else
+            catch (MsalException)
             {
-                return await GetAccessTokenOnBehalfOfUser(application, context.User, scopes, tenant);
-            }
-        }
+                string authorizationHeader = context.Request.Headers["Authorization"];
 
-        /// <summary>
-        /// In a Web API, adds to the MSAL.NET cache, the account of the user for which a bearer token was received when the Web API was called.
-        /// An access token and a refresh token are added to the cache, so that they can then be used to acquire another token on-behalf-of the
-        /// same user in order to call to downstream APIs.
-        /// </summary>
-        /// <param name="tokenValidationContext">Token validation context passed to the handler of the OnTokenValidated event
-        /// for the JwtBearer middleware</param>
-        /// <param name="scopes">[Optional] scopes to pre-request for a downstream API</param>
-        /// <example>
-        /// From the configuration of the Authentication of the ASP.NET Core Web API (for example in the Startup.cs file)
-        /// <code>JwtBearerOptions option;</code>
-        ///
-        /// Subscribe to the token validated event:
-        /// <code>
-        /// options.Events = new JwtBearerEvents();
-        /// options.Events.OnTokenValidated = async context =>
-        /// {
-        ///   var tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();
-        ///   tokenAcquisition.AddAccountToCacheFromJwt(context);
-        /// };
-        /// </code>
-        /// </example>
-        public void AddAccountToCacheFromJwt(Microsoft.AspNetCore.Authentication.JwtBearer.TokenValidatedContext tokenValidatedContext,
-            IEnumerable<string> scopes)
-        {
-            if (tokenValidatedContext == null)
-                throw new ArgumentNullException(nameof(tokenValidatedContext));
+                if (string.IsNullOrEmpty(authorizationHeader))
```

Comment:
The method is also used on Web App, for which I don't think there will be an authorization header => rather have a if (then use the header) else (then use ClaimsPrincipal as a key)

Reply:
It can't be used on a WebApp because the method is called GetATOnBehalfOfUser(), so this method encapsulates the OBO grant and can be used only by the Web Api

---
In reply to: [318913909](https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/pull/93#discussion_r318913909) [](ancestors = 318913909)

Reply:
@jennyf19 : in fact no. Maybe that's stupid to have named the method of ITokenAcquisition this way, but this is THE method that enables either a Web App or a Web API to get a token when a user is involved from an ASP.NET Core Controller: in other words, it's not important for the developer of a controller action to understand if they are in a Web app or Web API (ASP.NET Core does all it can to minimize the differences), what is important is that the token is not for the app itsefl, but for the signed-in user (web app), or the user for which the API was called (Web API).

We can revisit the naming if you think that's a mistake, but if you look in web app samples, you'll see that the same meethod is used: https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/0ba646fd2beac5016e8750e24e576419b2301f25/2-WebApp-graph-user/2-1-Call-MSGraph/Controllers/HomeController.cs#L64

Again the Microsoft.Identity.Web project is reusable, and, in general synched between the tutorials (it's not at the moment because I want to merge this PR before re-bringing the latest refactoring done in https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2)